### PR TITLE
feat: Add CUDA Graph configuration support to MegatronPolicyWorker

### PR DESCRIPTION
**Pull request by sahgerlad**
# What does this PR do ?

Add CUDA Graph configuration support to MegatronPolicyWorker

# Issues

N/A - Feature addition

# Usage

```yaml
# In config YAML, under megatron_cfg:
megatron_cfg:
  enabled: true
  enable_cuda_graph: true           # Enable CUDA graph capture
  cuda_graph_scope: "attn"    # Optional: set the scope for CUDA graphs
  # ... other megatron config options
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

N/A

## Summary by CodeRabbit

* **New Features**
  * Added CUDA Graph configuration support for Megatron policy worker, allowing users to enable CUDA graphs and specify their scope (e.g., full_model).
  * Integrated RNG configuration that automatically enables RNG tracking when CUDA graphs are activated.

* **Tests**
  * Added comprehensive test coverage for CUDA graph configuration parsing with various parameter combinations.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: nemo_rl/models/megatron/setup.py [hunk 1]
```diff
@@ -481,6 +482,21 @@ def _apply_performance_config(model_cfg: Any, config: PolicyConfig) -> None:
                 f"Unsupported {type(attention_backend)=}, expected str or int"
             )
 
+    # CUDA Graph configuration
+    if "enable_cuda_graph" in config["megatron_cfg"]:
+        model_cfg.enable_cuda_graph = config["megatron_cfg"]["enable_cuda_graph"]
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py:487]: can you add these configs to https://github.com/NVIDIA-NeMo/RL/blob/9feb4b0f13e399cbcee3e7150724deaec44f280e/nemo_rl/models/policy/__init__.py#L177

**Review by terrykong — COMMENTED** [review]
@guyueh1 @ananthsub can you review?

File: nemo_rl/models/megatron/setup.py (hunk 1 above)
Comment (sahgerlad) [reply nemo_rl/models/megatron/setup.py:487]: Added `Literal` definition using reference from [Megatron-Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/scripts/performance/argument_parser.py#L28)

**Review by guyueh1 — APPROVED** [review]
LGTM; not necessarily have to be noted in PR, but I wonder what recipes can use it. IIUC it can only be used in non-colocated, and the full_iteration only works for dense model?

File: nemo_rl/models/policy/__init__.py [hunk 2]
```diff
@@ -224,6 +228,8 @@ class MegatronConfig(TypedDict):
     # Number of tokens per chunk when computing the fused linear CE loss.
     # Smaller values reduce peak memory further but may decrease throughput.
     linear_ce_fusion_chunk_size: NotRequired[int]
+    enable_cuda_graph: bool
+    cuda_graph_scope: CudaGraphScope
```
Comment (terrykong) [inline nemo_rl/models/policy/__init__.py:232]: [`__init__.py:231-232`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/policy/__init__.py#L231-L232)

**[BUG — blocking]** These fields are declared as `Required` but [`setup.py:486`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/megatron/setup.py#L486) treats them as optional (`if "enable_cuda_graph" in config[...]`).

This will break [`test_all_config_files_have_required_keys`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/tests/unit/test_config_validation.py#L96) for all ~160 example YAMLs that don't include these keys — that test uses Pydantic `TypeAdapter` which enforces `Required` vs `NotRequired` on TypedDicts.

```suggestion
    enable_cuda_graph: NotRequired[bool]
    cuda_graph_scope: NotRequired[CudaGraphScope]
```

File: nemo_rl/models/megatron/setup.py [hunk 3]
```diff
@@ -481,6 +482,28 @@ def _apply_performance_config(model_cfg: Any, config: PolicyConfig) -> None:
...
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py:487]: [`setup.py:486-487`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/megatron/setup.py#L486-L487)

**[BUG — blocking]** `TransformerConfig.enable_cuda_graph` is [explicitly deprecated](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L766-L770) in Megatron-LM, replaced by [`cuda_graph_impl`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L791-L797). At runtime, [`__post_init__`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L1696-L1698) emits a DeprecationWarning and [asserts `cuda_graph_impl == "none"`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L1688-L1691) before migrating.

Megatron-Bridge already uses the non-deprecated API — see [`_set_cuda_graph_overrides`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/f91542b90908ad08b7e13672feea03e27bedee27/scripts/performance/utils/overrides.py#L102-L124):

```python
# How Megatron-Bridge does it (overrides.py:108-113):
if cuda_graph_impl is not None:
    recipe.model.cuda_graph_impl = cuda_graph_impl
    if cuda_graph_impl != "none":
        recipe.rng.te_rng_tracker = recipe.model.use_te_rng_tracker = True
    else:
        recipe.rng.te_rng_tracker = recipe.model.use_te_rng_tracker = False
```

Recommend following this same pattern: replace `enable_cuda_graph: bool` with `cuda_graph_impl: str` (values: `"none"`, `"local"`, `"transformer_engine"`) and set `model_cfg.cuda_graph_impl` directly.

File: tests/unit/models/policy/test_megatron_worker.py [hunk 4]
```diff
@@ -2686,3 +2690,37 @@ def test_megatron_policy_flops_range_check(tiny_llama_model_path):
     finally:
         policy.shutdown()
         cluster.shutdown()
+
+
+def test_cuda_graph_config_parsing():
```
Comment (terrykong) [inline tests/unit/models/policy/test_megatron_worker.py:2695]: [`test_megatron_worker.py:2695-2726`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/tests/unit/models/policy/test_megatron_worker.py#L2695-L2726)

**[TEST — nit]** The test verifies the config helper builds the dict correctly, which is useful. One thing to consider: it might also be worth adding a case for an invalid `cuda_graph_scope` value to make sure the [`ValueError` at setup.py:492`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/megatron/setup.py#L491-L495) is exercised — but that's optional and could come in a follow-up.

File: nemo_rl/models/megatron/setup.py [hunk 5]
```diff
@@ -481,6 +482,28 @@ def _apply_performance_config(model_cfg: Any, config: PolicyConfig) -> None:
...
+        if "cuda_graph_scope" in config["megatron_cfg"]:
+            scope = config["megatron_cfg"]["cuda_graph_scope"]
+            valid_scopes = get_args(CudaGraphScope)
+            if scope not in valid_scopes:
+                raise ValueError(
+                    f"Invalid cuda_graph_scope '{scope}'. "
+                    f"Valid options are: {valid_scopes}"
+                )
+            model_cfg.cuda_graph_scope = scope
+            if not model_cfg.enable_cuda_graph:
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py:497]: [`setup.py:496-501`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/megatron/setup.py#L496-L501)

**[BUG]** Warning says "cuda_graph_scope will have no effect" but [line 496](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/megatron/setup.py#L496) already assigned `model_cfg.cuda_graph_scope = scope`. If `cuda_graph_impl` is non-"none" from another source (Megatron supports [`cuda_graph_impl: "local" | "transformer_engine"`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L791-L797)), the scope WILL take effect despite the warning. Either skip the assignment when disabled, or correct the warning.

File: nemo_rl/models/policy/__init__.py [hunk 6]
```diff
@@ -12,10 +12,14 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any, Literal, NotRequired, TypedDict, Union
+from typing import Any, Literal, NotRequired, TypedDict
 
 from nemo_rl.models.generation.interfaces import GenerationConfig
 
+CudaGraphScope = Literal[
+    "full_iteration", "attn", "mlp", "moe", "moe_router", "moe_preprocess", "mamba"
```
Comment (terrykong) [inline nemo_rl/models/policy/__init__.py:20]: [`__init__.py:19-21`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/policy/__init__.py#L19-L21)

**[BUG]** This `Literal` is brittle — it's already missing [`full_iteration_inference`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/enums.py#L80) and can't express [comma-separated compound scopes](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L1716) like `"attn,mlp"`. It will go out of sync whenever Megatron-LM updates the enum.

**Recommendation**: Drop the `CudaGraphScope` type alias. Declare both fields as `str` in the TypedDict with inline comments pointing to the upstream source of truth. Following the [Megatron-Bridge pattern](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/f91542b90908ad08b7e13672feea03e27bedee27/scripts/performance/utils/overrides.py#L102-L124), the fields should be:

```python
    # CUDA graph implementation. Valid values: "none", "local", "transformer_engine".
    # See: https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L791-L797
    cuda_graph_impl: NotRequired[str]
    # CUDA graph capture scope. Supports single values (e.g. "attn", "full_iteration")
    # or comma-separated compounds (e.g. "attn,mlp"). Scope validation depends on impl.
    # See: https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/enums.py#L70-L80
    cuda_graph_scope: NotRequired[str]
```

The `get_args(CudaGraphScope)` validation at [`setup.py:490`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/megatron/setup.py#L490) can then be removed — Megatron's own [`__post_init__`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L1706-L1732) already validates and normalizes scopes.

More broadly: new config fields that proxy upstream Megatron args should have inline comments referencing the upstream docs for full semantics and valid values (similar to [`defer_fp32_logits`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/policy/__init__.py#L198-L200) and [`use_linear_ce_fusion_loss`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/policy/__init__.py#L222-L227)).

File: nemo_rl/models/megatron/setup.py [hunk 7]
```diff
@@ -481,6 +482,28 @@ def _apply_performance_config(model_cfg: Any, config: PolicyConfig) -> None:
...
+                warnings.warn(
+                    "cuda_graph_scope is configured but enable_cuda_graph is False. "
+                    "The cuda_graph_scope setting will have no effect."
+                )
+        if model_cfg.enable_cuda_graph:
+            model_cfg.use_te_rng_tracker = True
+        else:
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py:504]: [`setup.py:502-505`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/nemo_rl/models/megatron/setup.py#L502-L505)

**[BUG]** Setting `enable_cuda_graph=False` relies on the deprecated flag to disable CUDA graphs. In Megatron-LM, the actual runtime check is [`cuda_graph_impl != "none"`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L1734) — this gates all CUDA graph setup, scope validation, and layer wrapping. The deprecated `enable_cuda_graph` is only consulted in [`__post_init__`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L1688-L1698) to migrate to `cuda_graph_impl="local"`.

Currently this works because nemo-rl builds `model_cfg` from scratch (so `cuda_graph_impl` starts at its [default `"none"`](https://github.com/NVIDIA/Megatron-LM/blob/23dd639cf3de30f3b9d8d0fae71ee31180be9ddd/megatron/core/transformer/transformer_config.py#L791)). But it's fragile — if `cuda_graph_impl` is ever added to the nemo-rl config surface or set via a checkpoint, setting `enable_cuda_graph=False` would not override it. This is another reason to migrate to `cuda_graph_impl` directly.

**Review by terrykong — COMMENTED** [review]
## Review Summary

> **⚠️ Merge conflicts detected** — this PR has conflicts with `main`. Please rebase and resolve before further review.

This PR adds CUDA Graph configuration support — a valuable feature. However, there are **2 blocking issues** and **6 suggestions** that should be addressed before merge.

### Blocking
1. **TypedDict fields must be `NotRequired`** — `enable_cuda_graph` and `cuda_graph_scope` are declared Required, breaking ~160 config validation tests
2. **`enable_cuda_graph` is deprecated in Megatron-LM** — replaced by `cuda_graph_impl` (emits DeprecationWarning, risks assertion crash)

### Performance Evidence
Since this feature affects training performance, it would be helpful to include some benchmark numbers in the PR description (e.g. tokens/sec or step time with and without CUDA graphs, on a representative config). This helps reviewers and future users understand the expected benefit and any trade-offs. The [`.coderabbit.yaml` pre-merge check](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/.coderabbit.yaml#L69-L75) also flags this.

### Additional: Example YAMLs
Per [`docs/design-docs/design-and-philosophy.md`](https://github.com/NVIDIA-NeMo/RL/blob/d9a8778e8b73980c6e257c3f637cf1cddb7ad6ac/docs/design-docs/design-and-philosophy.md#L134), exemplar YAMLs under `examples/configs/` should document new config keys. No example YAML includes `enable_cuda_graph` or `cuda_graph_scope`. Consider adding to representative Megatron configs (e.g. `grpo_math_1B_megatron.yaml`) with commented defaults.

### Devil's Advocate
3 findings confirmed | 3 downgraded | 1 disputed | PR value: confirmed (real feature, not trivial)

_Generated by Claude Code_

File: nemo_rl/models/megatron/setup.py (hunk 3 above)
Comment (guyueh1) [reply nemo_rl/models/megatron/setup.py:487]: I agree, we should use `cuda_graph_impl` instead of `enabled_cuda_graph`

File: nemo_rl/models/policy/__init__.py (hunk 6 above)
Comment (guyueh1) [reply nemo_rl/models/policy/__init__.py:20]: we can simply use a str here, the mcore will validate this at some point, i think

**Review by guyueh1 — COMMENTED** [review]
Like the comment that Terry left: we should use cuda_graph_impl instead of enabled_cuda_graph
